File: 077_combinations/Solution1/Solution.go
package main

import "fmt"

func combine(n int, k int) [][]int {
    var results [][]int
    var path []int

    dfs(&results, path, 1, n, k)

    return results
}

func dfs(results *[][]int, path []int, start, n, k int) {
    if k == 0 {
        *results = append(*results, append([]int{}, path...))
        return
    }

    for i := start; i <= n; i++ {
        path = append(path, i)
        dfs(results, path, i + 1, n, k - 1)
        path = path[:len(path) - 1]
    }
}

func main() {
    results := combine(4, 2)

    for _, result := range results {
        fmt.Println(result)
    }
}

File: 077_combinations/Solution1/Solution.py
class Solution:
    def combine(self, n: int, k: int) -> list:
        if n == 0 or k == 0 or n < k: return []

        results = []
        path = []

        def dfs(start: int, n: int, k: int):
            if k == 0:
                results.append(path[:])
                return

            for i in range(start, n + 1):
                path.append(i)
                dfs(i + 1, n, k - 1)
                path.pop()

        dfs(1, n, k)

        return results


if __name__ == '__main__':
    print(Solution().combine(4, 2))

